docs/proposals/20191030-machine-health-checking.md

```diff
@@ -0,0 +1,211 @@
+
+
+### User Stories
+
+#### Story 1
+- As a user of a Workload Cluster I only care about my apps availability so I want my cluster infrastructure to be self healing and the nodes to be remediated transparently in the case of failures.
+
+#### Story 2
+- As an operator of a Management Cluster I want my machines to be self-healing and to be recreated resulting in a new healthy node in the case of matching my unhealthy criteria.
+
+
+### Implementation Details/Notes/Constraints
+
+#### MachineHealthCheck CRD:
+- Enable watching a group of machines (based on a label selector).
+- Enable defining an unhealthy node criteria (based on a list of node conditions).
+- Enable setting a threshold of unhealthy nodes. If the current number is at or above this threshold no further remediation will take place. This can be expressed as an int or as a percentage of the total targets in the pool.
+
+
+E.g:
+- I want my worker machines to be remediated when the backed node has `ready=false` or `ready=Unknown` condition for more than 5m.
+- I want remediation to temporary short-circuit if the 40% or more of the targets of this pool are unhealthy at the same time.
+
+```yaml
+apiVersion: healthchecking.cluster-api.io/v1alpha3
+kind: MachineHealthCheck
+metadata:
+  name: example
+  namespace: machine-api
+spec:
+  selector:
+    matchLabels:
+      role: worker
+  unhealthyConditions:
+  - type:    "Ready"
+    status:  "Unknown"
+    timeout: "5m"
+  - type:    "Ready"
+    status:  "False"
+    timeout: "5m"
+  maxUnhealthy: "40%"
+status:
+  currentHealthy: 5
+  expectedMachines: 5
+```
+
+#### MachineHealthCheck controller:
+Watch:
+- Watch machineHealthCheck resources
+- Watch machines and nodes with an event handler e.g controller runtime `EnqueueRequestsFromMapFunc` which returns machineHealthCheck resources.
+
+Reconcile:
+- Fetch all the machines targeted by the MachineHealthCheck and operate over machine/node targets. E.g:
+```
+type target struct {
+  Machine capi.Machine
+  Node    *corev1.Node
+  MHC     capi.MachineHealthCheck
+}
+```
+
+- Calculate the number of unhealthy targets.
+- Compare current number against `maxUnhealthy` threshold and temporary short circuits remediation if the threshold is met.
+- Trigger remediation for unhealthy targets i.e request machines for deletion.
+
+Out of band:
+- The owning controller e.g machineSet controller reconciles to meet number of replicas and start the process to bring up a new machine/node.
+- The machine controller drains the unhealthy node.
+- The machine controller provider deletes the instance.
+- The machine controller deletes the machine.
+
+![Machine health check](./images/machine-health-check/mhc.svg)
+
+### Risks and Mitigations
+
+## Alternatives
+
+Considered to bake this functionality into machineSets. This was discarded as different controllers than a machineSet could be owning the targeted machines. For those cases as a user you still want to benefit from automated node remediation.
```

Comment:
(Overall comment for the entire document) would you please break long lines/paragraphs at ~ 100 characters per line? Thanks!